### ISSUE TITLE
when being removed from a department, staffer is not unassigned from any shifts in that dept

### ISSUE THREAD

**Issue by binary1230** [magfest/ubersystem#1570]
This might not be a bug, it might be a feature : )

I noticed while looking at #1561 / #1474 that we don't seem to unassign people from shifts at certain times:
1) when we remove someone from being assigned to a dept, it doesn't look like we unassign them from any shifts they have signed up for in that department
2) when we remove someone from being trusted in that dept, we don't unassign them from any restricted shifts they have signed up for in that dept.

I will note that when we remove someone from being a staffer, they do get unassigned from all shifts.

Repro:
- assign someone to a dept (i.e. Arcade)
- sign them up for a shift in arcade
- unassign them from arcade

expected: they are unassigned from Arcade shifts
actual: they are still signed up for their arcade shift

----

This might be something we want to fix, or it might be intended behavior.  I'll let @EliAndrewC, @kitsuta  and @bds002 comment

At any rate I won't be trying to address this in #1561 since the behavior is the same as it was previously.

**Comment by kitsuta** [comment]
I would say this is undesired behavior. As a user I would definitely assume that removing someone from a department removes their shifts.

Of course if we implement this we do want a confirmation page for doing so, lest we get some headaches by accidental misclicks.

**Comment by EliAndrewC** [comment]
The reason for this "feature" is that when people show up to Stops on-site and say "I'd like to volunteer" we want to be able to assign them to arbitrary shifts without having to go in and assign them to each department where they're taking one or more shifts.  As such, we explicitly want people to be able to hold shifts in departments where they're not assigned.

With that being said, we could certainly do things like make that behavior an ``AT_OR_POST_CON``-only thing, etc.

**Comment by bds002** [comment]
Yeah, I am going to go with: That was a fly over mistake in a hot bedded
attempt to fix something else.

However a catch here would be - if removed from a department, we would
remove them from future shifts, we would still want to see what shifts they
worked up to that point. My thought is that someone got removed from a
department for (Insert issue here) we would still want to see their old
shifts and any comments left by the corresponding DH's.

Brent

**Comment by binary1230** [comment]
Ah @bds002 that's a good point, it would erase their shift history too, hand't thought about that.

**Comment by thaeli** [comment]
I would recommend a confirm screen AND asking "do you want to perform these cleanup actions (recommended, only click no if you know what you're doing)"

**Comment by bds002** [comment]
What if we update the code to add them to the department if we add them to
an open shift.

Stops has a page... (I can't find it at the moment) that we use on site..
that shows ever open and available shift at the moment - across all
departments and that is part of what we use for the STaff Support folks...
and new volunteers on site.

Brent

**Comment by kitsuta** [comment]
@thaeli The confirm screen I was thinking of would be similar to how the group delete confirm works. It would say something like "Are you sure you want to unassign this attendee from {DEPT}? Doing so will also remove the following shifts they are signed up for:

{list_of_shifts}

[Yes] [No]"

**Comment by EliAndrewC** [comment]
> What if we update the code to add them to the department if we add them to an open shift.

That does seem good, and is a lot easier to implement than the confirmation screen, though it would supplement rather than replace the confirmation screen.

**Comment by thaeli** [comment]
So are people able to be on a shift without being assigned to that department, or not? Either way the system needs to act consistently everywhere. So if there are ANY use cases where someone should have valid shifts in a department and not be assigned to that department, allow that across the board. If we are confident that there are no use cases requiring this, and since this is a core function I'd prefer that we go ahead and think about other events as well, then enforce the rule that shifts must be for departments you're assigned to including automatic adding of departments (with confirm screen) and so forth.

A common example that comes to mind is, a staffer or volunteer for dept X goes and helps out in dept Y for a shift. You want to track this, but they're not really a member of dept Y. How should that be reflected in the system?

**Comment by EliAndrewC** [comment]
There's definitely a use case for an admin being able to assign someone to a department which that person is not currently assigned to.  The question is whether we should either

- continue doing what we're doing now, and just allow that, with the implicit understanding that people can be assigned to shifts (by admins) in departments they're not in
- start having this action automatically assign them to that department, and then require that people belong to departments where they have shifts

or some combination/hybrid approach.  I'll note that the first option is less work, which is why we've done it that way in the past.

**Comment by bds002** [comment]
I think we should automatically add someone to the department (with a verification box if needed) if they work a shift in that department.  This allows the DH of that department to go in and rate the staffer for the shift, and if the staffers likes the department, they can easily pick up more shifts there. 

If we remove someone from a department which happens often, but not necessarily at event. That should not remove the shifts they have already worked in that department. 

As for Thaeli's case where someone is in X department and goes and works in Y department for 1 shift as a help case. - Unless they are coming out of Staff Support then the department they are assigned to would have to sign off on them helping somewhere else during shift time.  If it's outside of their shift hours.. I still think they should be added to the department they helped in. If it's a one off thing, we have the Non-shift hours that can used to track that as well.

In the same case above but staffer comes out of Staff Support - we should be recording hours in staff support, or if we actually assign them to open department based shift hours, then the DH should be signing of on hours.  (Fox, Caramon Stanley) is a good example of this from M12.  He came to stops and wanted to know where he could work.... instead of using him in Staff Support at the time, we assigned him to Open Console shifts later that day... he has been attached to Consoles ever since.
